Inserting consecutive <picture> elements displays the same image twice

pxlcoder opened the pull request:
#### afc67068fec68a95864c3a7c4b14d28c5dba277c
<pre>
Inserting consecutive &lt;picture&gt; elements displays the same image twice
<a href="https://bugs.webkit.org/show_bug.cgi?id=271715">https://bugs.webkit.org/show_bug.cgi?id=271715</a>
<a href="https://rdar.apple.com/123795045">rdar://123795045</a>

Reviewed by Ryosuke Niwa and Wenson Hsieh.

Currently, when inserting two successive elements of the form:

```
&lt;picture&gt;
    &lt;source&gt;
    &lt;img&gt;
&lt;/picture&gt;
```

the resulting markup is:

```
&lt;picture&gt;
    &lt;source&gt;
    &lt;img&gt;
    &lt;img&gt;
&lt;/picture&gt;
```

This results in two images displaying the same `&lt;source&gt;` content.

There are two distinct issues which result in the undesirable markup.

1. The second `&lt;picture&gt;` and `&lt;source&gt;` are dropped by `serializePreservingVisualAppearance`.
   When getting `VisiblePosition`s encompassing the inserted content, only the
   `&lt;img&gt;` is selected, as the first visible position inside the inserted element
   is before the `&lt;picture&gt;`. Consequently, the inserted `&lt;picture&gt;` and
   `&lt;source&gt;` are skipped when serializing nodes.

2. As the selection is inside the first `&lt;picture&gt;` after insertion, subsequent
   content also gets added to the same `&lt;picture&gt;` element.

To fix, `serializePreservingVisualAppearance` must preserve the `&lt;picture&gt;` and
`&lt;source&gt;`. Additionally, any insertions made while the selection is inside
`&lt;picture&gt;`, should be moved outside.

* LayoutTests/editing/inserting/insert-text-after-picture-expected.txt: Added.
* LayoutTests/editing/inserting/insert-text-after-picture.html: Added.
* LayoutTests/editing/pasteboard/copy-picture-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-picture.html: Added.
* Source/WebCore/editing/Editing.cpp:
(WebCore::canHaveChildrenForEditing):

Return false to ensure children added by editing appear before or after the
`&lt;picture&gt;`, rather than inside it.

`canContainRangeEndPoint` itself is not modified, as doing that introduces
selection and caret painting issues with &lt;picture&gt;. The &lt;img&gt; element itself
should remain selectable.

* Source/WebCore/editing/markup.cpp:
(WebCore::highestAncestorToWrapMarkup):

Return the `&lt;picture&gt;` element if it is a common ancestor of start/end, ensuring
`&lt;picture&gt;` is preserved in the serialized markup.

(WebCore::serializePreservingVisualAppearanceInternal):

Adjust the start Position, so that the `&lt;source&gt;` is included.

Canonical link: <a href="https://commits.webkit.org/276754@main">https://commits.webkit.org/276754@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fabcc7919197b232fc2831f7ed31008f32673a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24592 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48133 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41475 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47774 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21986 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37271 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46045 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21695 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39230 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18393 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19094 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40319 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3513 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41800 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40644 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49859 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20452 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44334 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39389 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43161 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10128 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22119 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21447 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/html/HTMLPictureElement.cpp:
```diff
@@ -75,5 +75,18 @@ bool HTMLPictureElement::isSystemPreviewImage()
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+bool HTMLPictureElement::isMultiRepresentationHEIC() const
+{
+    if (countChildNodes() != 2)
```
why 2?

pxlcoder replied:
`<source>` + `<img>`

rniwa replied:
Surely there could be more than one `source` element? You can specify multiple `source` with different media queries. We should add a test for that case as well.

pxlcoder replied:
Added a test case for multiple `source` with media queries.

rniwa commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -73,6 +74,11 @@ static bool isVisiblyAdjacent(const Position&, const Position&);
 
 bool canHaveChildrenForEditing(const Node& node)
 {
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
```
I don't think we should modifying this function.
We should instead override `HTMLPictureElement::canContainRangeEndPoint`.

pxlcoder replied:
It's the first thing I tried, but that causes many other editing issues. Including an inability to modify the selection using the arrow keys, and clicking around the element to modify the selection.

whsieh commented on Source/WebCore/html/HTMLPictureElement.cpp:
```diff
@@ -75,5 +75,18 @@ bool HTMLPictureElement::isSystemPreviewImage()
+    for (auto& element : childrenOfType<HTMLImageElement>(*this))
+        return element.isMultiRepresentationHEIC();
+
+    return false;
```
The immediate return from the `for` loop was a bit unexpected, though it looks right. I wonder if this might be clearer?

```suggestion
    RefPtr firstImage = childrenOfType<HTMLImageElement>(*this).first();
    return firstImage && firstImage->isMultiRepresentationHEIC();
```

pxlcoder committed:
Inserting consecutive <picture> elements displays the same image twice
https://bugs.webkit.org/show_bug.cgi?id=271715
rdar://123795045

Reviewed by Ryosuke Niwa and Wenson Hsieh.

Currently, when inserting two successive elements of the form:

```
<picture>
    <source>
    <img>
</picture>
```

the resulting markup is:

```
<picture>
    <source>
    <img>
    <img>
</picture>
```

This results in two images displaying the same `<source>` content.

There are two distinct issues which result in the undesirable markup.

1. The second `<picture>` and `<source>` are dropped by `serializePreservingVisualAppearance`.
   When getting `VisiblePosition`s encompassing the inserted content, only the
   `<img>` is selected, as the first visible position inside the inserted element
   is before the `<picture>`. Consequently, the inserted `<picture>` and
   `<source>` are skipped when serializing nodes.

2. As the selection is inside the first `<picture>` after insertion, subsequent
   content also gets added to the same `<picture>` element.

To fix, `serializePreservingVisualAppearance` must preserve the `<picture>` and
`<source>`. Additionally, any insertions made while the selection is inside
`<picture>`, should be moved outside.

* LayoutTests/editing/inserting/insert-text-after-picture-expected.txt: Added.
* LayoutTests/editing/inserting/insert-text-after-picture.html: Added.
* LayoutTests/editing/pasteboard/copy-picture-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-picture.html: Added.
* Source/WebCore/editing/Editing.cpp:
(WebCore::canHaveChildrenForEditing):

Return false to ensure children added by editing appear before or after the
`<picture>`, rather than inside it.

`canContainRangeEndPoint` itself is not modified, as doing that introduces
selection and caret painting issues with <picture>. The <img> element itself
should remain selectable.

* Source/WebCore/editing/markup.cpp:
(WebCore::highestAncestorToWrapMarkup):

Return the `<picture>` element if it is a common ancestor of start/end, ensuring
`<picture>` is preserved in the serialized markup.

(WebCore::serializePreservingVisualAppearanceInternal):

Adjust the start Position, so that the `<source>` is included.

Canonical link: https://commits.webkit.org/276754@main
